dig.go

```diff
@@ -0,0 +1,277 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package dig
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"reflect"
+)
+
+var (
+	_noValue reflect.Value
+	_errType = reflect.TypeOf((*error)(nil)).Elem()
+)
+
+// A Container is a directed, acyclic graph of dependencies. Dependencies are
+// constructed on-demand and returned from a cache thereafter, so they're
+// effectively singletons.
+type Container struct {
+	nodes map[reflect.Type]node
+	cache map[reflect.Type]reflect.Value
+}
+
+// New constructs a ready-to-use Container.
+func New() *Container {
+	return &Container{
+		nodes: make(map[reflect.Type]node),
+		cache: make(map[reflect.Type]reflect.Value),
+	}
+}
+
+// Provide teaches the Container how to construct one or more new types. It
+// accepts either a constructor function or an already-constructed object.
+//
+// Any function passed to Provide is assumed to be a constructor. Constructors
+// can take any number of parameters, which will be supplied by the Container
+// on demand. They must return at least one non-error value, all of which are
+// then available in the Container. If the last returned value is an error, the
+// Container inspects it to determine whether the constructor succeeded or
+// failed. Regardless of position, returned errors are never put into the
+// Container's dependency graph.
+//
+// All non-functions (including structs, pointers, Go's built-in collections,
+// and primitive types like ints) are inserted into the Container as-is.
+func (c *Container) Provide(constructor interface{}) error {
+	ctype := reflect.TypeOf(constructor)
+	if ctype == nil {
+		return errors.New("can't provide an untyped nil")
+	}
+	// Since we want to wrap any errors, don't return early.
+	var err error
+	if ctype.Kind() != reflect.Func {
+		err = c.provideInstance(constructor)
+	} else {
+		err = c.provideConstructor(constructor, ctype)
+	}
+
+	if err == nil {
+		return nil
+	}
+	return fmt.Errorf("can't provide %v: %v", ctype, err)
+}
+
+// Invoke runs a function, supplying its arguments from the Container. If the
+// function's last return value is an error, that error is propagated to the
+// caller. All other returned values (if any) are ignored.
```

Comment:
> We no longer Invoke the module functions, we only invoke the service code, which relies on types provided by the modules.

I don't think that would work in real world scenarios. Invoke doesn't have any mechanism to prevent its usage to not invoke module funcs. How do you prevent someone from writing a stack which provides module functions?